salt/states/boto_ec2.py

```diff
@@ -1240,3 +1240,205 @@ def volumes_tagged(name, tag_maps, authoritative=False, region=None, key=None,
         ret['comment'] = 'Error updating requested volume tags.'
         ret['result'] = False
     return ret
+
+
+def volume_present(name, volume_name=None, volume_id=None, instance_name=None,
+                   instance_id=None, device=None, size=None, snapshot_id=None,
+                   volume_type=None, iops=None, encrypted=False, kms_key_id=None,
+                   region=None, key=None, keyid=None, profile=None):
+    '''
+    Ensure the EC2 volume is present and attached.
+
+    ..
+
+    name
+        State definition name.
+
+    volume_name
+        The Name tag value for the volume. If no volume with that matching name tag is found,
+        a new volume will be created. If multiple volumes are matched, the state will fail.
+
+    volume_id
+        Resource ID of the volume. Exclusive with 'volume_name'.
+
+    instance_name
+        Attach volume to instance with this Name tag.
+        Exclusive with 'instance_id'.
+
+    instance_id
+        Attach volume to instance with this ID.
+        Exclusive with 'instance_name'.
+
+    device
+        The device on the instance through which the volume is exposed (e.g. /dev/sdh)
+
+    size
+        The size of the new volume, in GiB. If you're creating the volume from a snapshot
+        and don't specify a volume size, the default is the snapshot size. Optionally specified
+        at volume creation time; will be ignored afterward. Requires 'volume_name'.
+
+    snapshot_id
+        The snapshot ID from which the new Volume will be created. Optionally specified
+        at volume creation time; will be ignored afterward. Requires 'volume_name'.
+
+    volume_type
+        The type of the volume. Optionally specified at volume creation time; will be ignored afterward.
+        Requires 'volume_name'.
+        Valid volume types for AWS can be found here:
+        http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSVolumeTypes.html
+
+    iops
+        The provisioned IOPS you want to associate with this volume. Optionally specified
+        at volume creation time; will be ignored afterward. Requires 'volume_name'.
+
+    encrypted
+        Specifies whether the volume should be encrypted. Optionally specified
+        at volume creation time; will be ignored afterward. Requires 'volume_name'.
+
+    kms_key_id
+        If encrypted is True, this KMS Key ID may be specified to encrypt volume with this key.
+        Optionally specified at volume creation time; will be ignored afterward.
+        Requires 'volume_name'.
+        e.g.: arn:aws:kms:us-east-1:012345678910:key/abcd1234-a123-456a-a12b-a123b4cd56ef
+
```

Comment:
You'd also want to pass wait_for_sync through the state module, so folks can choose to wait or not.